File: .husky/pre-push.js
import { execSync } from 'child_process';

try {
  // Check if the latest commit message contains [skip-ci] or [skip ci]
  const latestCommitMessage = execSync('git log -1 --pretty=%B', { encoding: 'utf8' }).trim();

  if (latestCommitMessage.match(/\[(skip-ci|skip ci)\]/i)) {
    console.log('⏭️  Skipping tests due to [skip-ci] in commit message.');
    process.exit(0);
  }

  // Run smart tests first - fail fast if tests don't pass
  console.log('Running smart tests...');
  execSync('node tests/smart-test-runner.js', { stdio: 'inherit' });
  console.log('Tests passed.');
} catch (e) {
  console.error('Pre-push hook failed: Tests did not pass.', e.message);
  process.exit(1);
}

File: .husky/pre-push.py
#!/usr/bin/env python3
"""
Pre-push hook to run smart tests before pushing.
Cross-platform implementation using Python.
"""
import subprocess
import sys


def main():
    """Run smart tests before allowing push."""
    try:
        # Check if the latest commit message contains [skip-ci] or [skip ci]
        result = subprocess.run(
            ["git", "log", "-1", "--pretty=%B"],
            capture_output=True,
            text=True,
            check=True
        )

        latest_commit_message = result.stdout.strip()

        if "[skip-ci]" in latest_commit_message.lower() or "[skip ci]" in latest_commit_message.lower():
            print("⏭️  Skipping tests due to [skip-ci] in commit message.")
            sys.exit(0)

        # Run smart tests first - fail fast if tests don't pass
        print("Running smart tests...")
        subprocess.run(
            ["node", "tests/smart-test-runner.js"],
            check=True
        )
        print("Tests passed.")
        sys.exit(0)

    except subprocess.CalledProcessError as e:
        print(f"Pre-push hook failed: Tests did not pass.", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Unexpected error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
